### PR TITLE
Bug fixes & new/expanded test cases

### PR DESCRIPTION
**Pull request by a-alamoudi**
Related to issues found in https://github.com/molinch/ConfigurationSubstitutor/issues/11#issuecomment-1003237216
- [FIX] Applied fixes to `EnableSubstitutionsWithDelimitedFallbackDefaults` to match behaviour of `EnableSubstitutions` when handling non-resolvable variables and `null` substituted values.
- [TEST] Added and expanded a couple of test cases to ensure the latest fixes.

### CODE REVIEW

File: ConfigurationSubstitutor/ConfigurationSubstitutor.cs [hunk 1]
```diff
@@ -72,22 +72,21 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
...
                 {
                     var delimitedVals = capture.Value.Split(new[] { _fallbackDefaultValueDelimiter }, 2, StringSplitOptions.None);
-                    // in case delimiter doesn't match
-                    if (delimitedVals.Length < 2)
+                    // in case delimiter matches
+                    if (delimitedVals.Length == 2)
                     {
-                        return capture.Value;
+                        value = value.Replace(capture.Value, delimitedVals[0]);
+                        // if declared EV isn't resolvable, assign it to provided fallback default value
```
Comment (molinch) [inline ConfigurationSubstitutor/ConfigurationSubstitutor.cs]: Could you elaborate on what means EV?

File: ConfigurationSubstitutor/ConfigurationSubstitutor.cs [hunk 2]
```diff
@@ -72,22 +72,21 @@ private string ApplySubstitution(IConfiguration configuration, string value, Has
...
+                        if (String.IsNullOrEmpty(configuration[delimitedVals[0]]))
+                        {
+                            configuration[delimitedVals[0]] = delimitedVals[1];
```
Comment (molinch) [inline ConfigurationSubstitutor/ConfigurationSubstitutor.cs]: I know this was already added in former PR, but I somehow missed that.
Not a big fan of altering source configuration, if we can keep the function pure and pass the value downwards sounds simpler/better to reason about. What do you think?

**Review by molinch — CHANGES_REQUESTED** [review]

File: ConfigurationSubstitutor/ConfigurationSubstitutor.cs (hunk 1 above)
Comment (a-alamoudi) [reply ConfigurationSubstitutor/ConfigurationSubstitutor.cs]: Absolutely, 'EV' refers to an environment variable. This section deals with the two tokens produced from breaking the captured value by the specified delimiter. Once execution arrives at this code segment, we're sure the delimiter matches, and that the first part (i.e., `delimitedVals[0]`) is the environment variable (EV).
Let's take the following example:
```
"Foo" = "${Var1:default_val}"
```
Substituting `config["Foo"]` with the start mark specified as `${`, the end mark as `}` and delimiter as `:`; we'd end up with `Var1` as the EV.

**Comment by a-alamoudi** [comment]
https://github.com/molinch/ConfigurationSubstitutor/pull/13#discussion_r776949125 @molinch I took another crack at it and avoided altering the source configuration by using a local dictionary instead. It passes all of the existing tests and behaves just the same. Please review and let me know what you think.
Thanks, and Happy New Year! 🙂

File: ConfigurationSubstitutor/ConfigurationSubstitutor.cs (hunk 1 above)
Comment (molinch) [reply ConfigurationSubstitutor/ConfigurationSubstitutor.cs]: Thanks for the explanation

**Review by molinch — APPROVED** [review]
Only did minor changes, thanks a lot @a-alamoudi !

**Comment by molinch** [comment]
On nuget as 1.4.1, have a very nice 2022 year!

**Comment by a-alamoudi** [comment]
@molinch Awesome! Thanks a lot, you too 😊

**Pull request merged**